osquery/events/linux/bpf/filesystem.cpp

```diff
@@ -0,0 +1,153 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <osquery/events/linux/bpf/filesystem.h>
+#include <osquery/events/linux/bpf/uniquedir.h>
+
+#include <fcntl.h>
+
+namespace osquery {
+bool Filesystem::open(tob::utils::UniqueFd& fd,
+                      const std::string& path,
+                      int flags) const {
+  fd.reset(-1);
+
+  {
+    auto handle = ::open(path.c_str(), flags);
+    if (handle == -1) {
+      return false;
+    }
+
+    fd.reset(handle);
+  }
+
+  return true;
+}
+
+bool Filesystem::openAt(tob::utils::UniqueFd& fd,
+                        int dirfd,
+                        const std::string& path,
+                        int flags) const {
+  fd.reset(-1);
+
+  {
+    auto handle = openat(dirfd, path.c_str(), flags);
+    if (handle == -1) {
+      return false;
+    }
+
+    fd.reset(handle);
+  }
+
+  return true;
+}
+
+bool Filesystem::readLinkAt(std::string& destination,
+                            int dirfd,
+                            const std::string& path) const {
+  std::vector<char> buffer(4096);
+
+  auto bytes_read =
+      readlinkat(dirfd, path.c_str(), buffer.data(), buffer.size() - 1U);
+
+  if (bytes_read == -1) {
+    return false;
+  }
+
+  destination.assign(buffer.data(), static_cast<std::size_t>(bytes_read));
+  return true;
+}
+
+bool Filesystem::read(std::vector<char>& buffer,
+                      int fd,
+                      std::size_t max_size) const {
+  buffer = {};
+
+  std::vector<char> read_buffer(4096);
+  while (buffer.size() < max_size) {
+    auto err = ::read(fd, read_buffer.data(), read_buffer.size());
+    if (err == 0) {
+      break;
+
+    } else if (err == -1) {
+      return false;
+    }
+
+    auto destination_offset = buffer.size();
+
+    auto bytes_read = static_cast<std::size_t>(err);
+    buffer.resize(buffer.size() + bytes_read);
+
+    std::memcpy(
+        buffer.data() + destination_offset, read_buffer.data(), bytes_read);
+  }
+
+  if (buffer.empty()) {
+    return false;
+  }
+
+  return true;
+}
+
+bool Filesystem::enumFiles(int dirfd, EnumFilesCallback callback) const {
```

Comment:
Check if `std::filesystem` has an implementation of this directory walk that you can reuse.

Reply:
I am using fd-based functions to reduce race conditions when traversing paths; the std::filesystem implementation would not support this